session-helper/background-monitor.c

```diff
@@ -28,12 +28,86 @@
 #include <libnotify/notify.h>
 #include "background-monitor.h"
 #include "app/flatpak-permission-dbus-generated.h"
+#include "app/gnome-shell-introspect-dbus-generated.h"
 
 #define PERMISSION_TABLE "background"
 #define PERMISSION_ID "background"
 
 typedef enum { UNSET, NO, YES, ASK } Permission;
 
+static GnomeShellIntrospect *shell = NULL;
+
+static void
+init_gnome_shell (GDBusConnection *connection)
+{
+  g_autoptr(GError) error = NULL;
+
+  shell = gnome_shell_introspect_proxy_new_sync (connection,
+                                                 G_DBUS_PROXY_FLAGS_NONE,
+                                                 "org.gnome.Shell",
+                                                 "/org/gnome/Shell/Introspect",
+                                                 NULL, &error);
+  if (shell == NULL)
+    g_debug ("No gnome-shell introspection: %s", error->message);
+}
+
+static GnomeShellIntrospect *
+get_gnome_shell (void)
+{
+  return shell;
+}
+
+typedef enum { BACKGROUND, RUNNING, ACTIVE } AppState;
+
+static GHashTable *
+get_app_states (void)
+{
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GVariant) windows = NULL;
+  g_autoptr(GHashTable) app_states = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+
+  if (!gnome_shell_introspect_call_get_windows_sync (get_gnome_shell (), &windows, NULL, &error))
+    g_debug ("Could not get window list: %s", error->message);
+
+  if (windows)
```

Comment:
Jonas pointed out that the applications list only includes apps with open windows, so the filtering we do here is not necessary.